### ISSUE TITLE
Ensure outputs are downloaded before sending TargetComplete event via BEP

### ISSUE THREAD

**Issue by coeuvre** [bazelbuild/bazel#17798]
### Description of the feature request:

Currently, when building with `--remote_download_toplevel`, outputs of toplevel targets / aspects are only guaranteed downloaded before the build command exits.

However, the consumers of BEP, e.g. IDE, may listen to `TargetComplete` event from BEP and expect outputs are ready to be processed when the event is fired.

### What underlying problem are you trying to solve with this feature?

We should wait for the downloads before sending `TargetComplete` event to not break the expectations for consumers of BEP.

### Any other information, logs, or outputs that you want to share?

#6862
b/273514526

**Comment by clint-stripe** [comment]
@coeuvre is there any work around besides disabling `--remote_download_toplevel`?

We have been using that flag for a while, and we're blocked by this issue as we upgrade from 5.4.0 to 6.1.1.

Our CI system uses BEP events to report test failures early and to locate artifacts -- we're seeing a `last_message` BEP event before some artifacts are written to disk, which is breaking assumptions our system made about locating build outputs on disk.

**Comment by coeuvre** [comment]
I am not aware of a workaround, but I am working on the fix and hopefully include it in 6.2.

**Comment by coeuvre** [comment]
@bazel-io flag

**Comment by keertk** [comment]
@bazel-io fork 6.2.0

**Comment by coeuvre** [comment]
Fixed by https://github.com/bazelbuild/bazel/commit/a5dde12b0d945b54967969fafa4d5e40692f9a44.

**Comment by keertk** [comment]
@coeuvre should this be included in 6.3.0?

**Comment by keertk** [comment]
@bazel-io fork 6.3.0

**Comment by keertk** [comment]
Hi @coeuvre wanted to check if we should hold off on 6.3 for this? We tentatively scheduled RC1 for 7/5.

**Comment by coeuvre** [comment]
Yes, we should. I will work on the workaround for 6.3.0 now.

**Comment by hoj-stripe** [comment]
@coeuvre Hi! We're trying to use the `--experimental_remote_cache_eviction_retries` flag and start evicting remote cache, but we're getting errors since we're also using`--experimental_remote_download_outputs=toplevel`. After bisecting on commits, we found out that the exact issue we are facing is unblocked with this commit(https://github.com/bazelbuild/bazel/commit/a5dde12b0d945b54967969fafa4d5e40692f9a44) mentioned in this issue.

I was wondering if you could help us unblock by backporting this commit onto 6.x. Have you had any luck on cherrypicking this commit to Bazel 6.x or finding the workaround? Any help would be appreciated!

**Comment by coeuvre** [comment]
https://github.com/bazelbuild/bazel/commit/a5dde12b0d945b54967969fafa4d5e40692f9a44 cannot be cherry-picked into 6.x because of missing many other commits. The [dedicated fix](https://github.com/bazelbuild/bazel/pull/18815) I created for 6.3 is a subset of https://github.com/bazelbuild/bazel/commit/a5dde12b0d945b54967969fafa4d5e40692f9a44.

Can you share more details about the error so that I can learn the underlying issue maybe create another dedicated fix? (it's better we can move the discussion into a new issue)
